### PR TITLE
Fix 28752 part 4

### LINKED ISSUE #28752: Document syntax restrictions for functional `id`s
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector

### What specific section or headline is this issue about?

Syntax -> Exceptions (or wherever the information should go)

### What information was incorrect, unhelpful, or incomplete?

A contributor on translated-content (@AppyGG) asked about adding an example to the `querySelector` page to explain how to deal with CSS `id`s that start with a number.

### What did you expect to see?

If possible, a note, example, or some sort of mention regarding this special case. Below is the example proposed by @AppyGG in mdn/translated-content#14872, though I think it could also be helpful to give an example of how to escape `id`s starting with a number as explained in [this StackOverflow post/answer](https://stackoverflow.com/questions/20306204/using-queryselector-with-ids-that-are-numbers):

> **Note :** If the selector selects an `id` and this `id` starts with a number (0-9), a syntax error is returned.

```html
<div id="0123"></div>
<div id="abcd"></div>

<script>
  document.querySelector("#0123"); // syntax error
  document.querySelector("#abcd"); // selects the second div
</script>
```

### Do you have any supporting links, references, or citations?

From the translated-content PR:

- [StackOverflow discussion about this problem](https://stackoverflow.com/questions/20306204/using-queryselector-with-ids-that-are-numbers)
- [W3Docs about HTML5 IDs](https://www.w3docs.com/snippets/html/which-are-the-valid-values-for-the-id-attribute.html#:~:text=In%20HTML5%2C%20the%20id%20attribute,id%20attribute%20is%20case%20sensitive.)

### Do you have anything more you want to share?

cf. mdn/translated-content#14872 for the related pull request.

### LINKED ISSUE #34253: What characters are allowed in a class name?
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/class

### What specific section or headline is this issue about?

The main discription of the class attribute

### What information was incorrect, unhelpful, or incomplete?

I think that "space-separated list" is not a clear enough description of what the value can be. Can it include any non-space unicode character? There is a link to the specification, but it is necessary to follow a second link after that to find this information.

### What did you expect to see?

The actual rules, plus some guidance on best practices.
For example:
A space-separated list of tokens, each consisting of any character other than a space. For maximum compatibility, it may be recommended to use only letters, numbers, dashes and underscores.

### Do you have any supporting links, references, or citations?

Link in article to spec: https://html.spec.whatwg.org/multipage/dom.html#global-attributes:classes-2
Extra links found in the spec, when read together gives a more complete answer
https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#set-of-space-separated-tokens
https://infra.spec.whatwg.org/#ascii-whitespace

### Do you have anything more you want to share?

I encountered the class name "N/A" in the wild, and I came to MDN to find out if it is a valid class name.

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/web/html/global_attributes/class`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/class
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/html/global_attributes/class/index.md
* Last commit: https://github.com/mdn/content/commit/ba96f2f183353872db6d9242c7d2dffe2dbc0c35
* Document last modified: 2023-02-24T05:52:35.000Z

</details>

### PR DESCRIPTION
**Pull request by wbamberg**
This is part 4 of the fix for https://github.com/mdn/content/issues/28752, as described in https://github.com/mdn/content/issues/28752#issuecomment-1707035191.

It follows on from https://github.com/mdn/content/pull/34340, https://github.com/mdn/content/pull/34601, and https://github.com/mdn/content/pull/35110, and updates the `class` and `id` attribute pages to note the issues that can arise when these attribute values are not valid CSS identifiers.

- fixes https://github.com/mdn/content/issues/28752
- fixes https://github.com/mdn/content/issues/34253

### CODE REVIEW

File: files/en-us/web/html/global_attributes/class/index.md [hunk 1]
```diff
@@ -7,12 +7,24 @@ browser-compat: html.global_attributes.class
...
 
+### Syntax
+
+The `class` attribute is a list of class values separated by [ACSII whitespace](/en-US/docs/Glossary/Whitespace#in_html).
+
+Each class value may contain any Unicode characters (except, of course, ASCII whitespace). However, when used in CSS selectors, either from JavaScript using APIs like {{domxref("Document.querySelector()")}} or in CSS stylesheets, class attribute values must be valid [CSS identifiers](/en-US/docs/Web/CSS/ident). This means that if a class attribute value is not a valid CSS identifier (for example, `my?class` or `1234`) then it must be escaped before being used in a selector, either using the {{domxref("CSS.escape_static", "CSS.escape()")}} method or [manually](/en-US/docs/Web/CSS/ident#escaping_characters).
+
+For this reason, it's recommended that developers choose values for class attributes that are valid CSS identifiers.
```
Comment (Josh-Cena) [inline files/en-us/web/html/global_attributes/class/index.md:26]:
```suggestion
For this reason, it's recommended that developers choose values for class attributes that are valid CSS identifiers that don't require escaping.
```

File: files/en-us/web/html/global_attributes/id/index.md [hunk 2]
```diff
@@ -7,25 +7,35 @@ browser-compat: html.global_attributes.id
...
+const content = window.preamble.textContent;
+```
+
+### Syntax
+
+An ID attribute's value must not contain [ACSII whitespace](/en-US/docs/Glossary/Whitespace#in_html) characters. Browsers treat non-conforming IDs that contain whitespace as if the whitespace is part of the ID. In contrast to the [`class`](/en-US/docs/Web/HTML/Global_attributes#class) attribute, which allows space-separated values, elements can only have one single ID value.
+
+Technically, the value for an ID attribute may contain any other Unicode character. However, when used in CSS selectors, either from JavaScript using APIs like {{domxref("Document.querySelector()")}} or in CSS stylesheets, ID attribute values must be valid [CSS identifiers](/en-US/docs/Web/CSS/ident). This means that if an ID attribute value is not a valid CSS identifier (for example, `my?id` or `1234`) then it must be escaped before being used in a selector, either using the {{domxref("CSS.escape_static", "CSS.escape()")}} method or [manually](/en-US/docs/Web/CSS/ident#escaping_characters).
+
+For this reason, it's recommended that developers choose values for ID attributes that are valid CSS identifiers.
```
Comment (Josh-Cena) [inline files/en-us/web/html/global_attributes/id/index.md:36]:
```suggestion
For this reason, it's recommended that developers choose values for ID attributes that are valid CSS identifiers that don't require escaping.
```

**Review by Josh-Cena — APPROVED** [review]
Very nice!

File: files/en-us/web/html/global_attributes/class/index.md [hunk 3]
```diff
@@ -7,12 +7,24 @@ browser-compat: html.global_attributes.class
...
+The **`class`** [global attribute](/en-US/docs/Web/HTML/Global_attributes) is a list of the classes of the element, separated by [ACSII whitespace](/en-US/docs/Glossary/Whitespace#in_html).
 
 {{EmbedInteractiveExample("pages/tabbed/attribute-class.html","tabbed-standard")}}
 
+## Description
+
+Classes allow CSS and JavaScript to select and access specific elements via the [class selectors](/en-US/docs/Web/CSS/Class_selectors) or functions like the {{domxref("document.getElementsByClassName()")}}.
+
 Though the specification doesn't put requirements on the name of classes, web developers are encouraged to use names that describe the semantic purpose of the element, rather than the presentation of the element. For example, _attribute_ to describe an attribute rather than _italics_, although an element of this class may be presented by _italics_. Semantic names remain logical even if the presentation of the page changes.
```
Comment (Josh-Cena) [inline files/en-us/web/html/global_attributes/class/index.md:18]: Not part of the PR but I suggest removing this paragraph. This is false for many styling frameworks which use utility classes describing the presentation. Afterwards this section may be too thin and you may consider moving the first paragraph back to the intro. :)

Comment (wbamberg) [reply files/en-us/web/html/global_attributes/class/index.md:18]: I opted not to do this, in this PR anyway, because this is [directly from the spec](https://html.spec.whatwg.org/multipage/dom.html#global-attributes:classes-2). So although I'm not entirely against removing this paragraph, I think the question is potentially controversial enough for me not to want to tie this very long-running bug fix to it :).
